Elder friday night PR

fiddlemath opened the pull request:
We currently have a heavily-edited, many-eyeballs edition of the code. It's ripe for a full PR and commit to `reviewed`!

tbrent committed:
init implementation. compiles, but untested

tbrent committed:
reorder and commenting

tbrent committed:
refactor onlyWhitelist

tbrent committed:
add rebalance function signatures

tbrent committed:
add nonce

tbrent committed:
WIP still, but checkpoint anyway

tbrent committed:
factory design init checkpoint

tbrent committed:
rename functions and switch order of emits

tbrent committed:
separate accepting from enacting and add 12 hour delay

tbrent committed:
add start of extra maxIssuance logic

tbrent committed:
finish issueMax and redeemMax

tbrent committed:
dual-proposal design compiles

tbrent committed:
comments and organization

tbrent committed:
remove openzeppelin Ownable

tbrent committed:
stash for rename

tbrent committed:
Update README.md

tbrent committed:
init abi and artifacts

tbrent committed:
init reserve + eternal storage

tbrent committed:
tests passing on initial setup

tbrent committed:
top-level main contract structure

tbrent committed:
add max supply to RSV and move ownership out

tbrent committed:
README

tbrent committed:
remove abi, artifact, and soltools

tbrent committed:
move contracts out of folders, add make abi

tbrent committed:
get tests back to working

tbrent committed:
use openzeppelin IERC20 instead of in-file

fiddlemath committed:
Create an abi directory if needed, during ABI generation.

tbrent committed:
make contracts + customize OptimizeRuns per target

tbrent committed:
checkpoint: deployment succeeding with commented out code

tbrent committed:
reduce size of require messages

tbrent committed:
refactor manager auth functions for succintness

tbrent committed:
switch all modifiers to using internal functions

tbrent committed:
fix reserve tests and start manager tests

tbrent committed:
almost there! only one more function to uncomment

tbrent committed:
finalize manager under the bytecode limit

tbrent committed:
use 100 max tokens

tbrent committed:
switch proposal status to enum

tbrent committed:
fix missing status state change

tbrent committed:
add changeOwner to vault + reserve. passing manager deploy tesT

tbrent committed:
manager constructor unittest

tbrent committed:
refactor contracts into folders

tbrent committed:
move proposal events into manager

tbrent committed:
remove unnecessary max functions

fiddlemath committed:
clean target for Makefile

fiddlemath committed:
Change "abi" generation to output solc combined-json files into evm/

This version continues also to emit the JSON format expected by sol-coverage; those
now get made in sol-coverage-evm/

fiddlemath committed:
trivial conflict resolution

fiddlemath committed:
Added current draft of design docs, with use cases and security properties.. So drafty!

tbrent committed:
add requireTxStrongly/Weakly

fiddlemath committed:
Added detailed notes on deployment.

fiddlemath committed:
Set gitattributes so that Solidity files are syntax-highlighted on github.

tbrent committed:
add basket swap

tbrent committed:
remove changeOwner from Reserve

tbrent committed:
add auth protection to executeProposal

tbrent committed:
fix basket swap ordering bug in executeProposal

tbrent committed:
test for proposeNewBasket

tbrent committed:
fix basket bugs and simplify

tbrent committed:
safemath for proposal math

tbrent committed:
make manager fully collateralized check public

tbrent committed:
fix Ownable error message

tbrent committed:
explain modifier approach in comments

tbrent committed:
add missing require

tbrent committed:
make delay an auth parameter

tbrent committed:
test updates for easier testing of proposal refactor

tbrent committed:
make auth control uniform

tbrent committed:
add BasketChanged event

tbrent committed:
add ManagerTransferred event to vault constructor

fiddlemath committed:
Merging.

tbrent committed:
change vault events

tbrent committed:
add vault tests

tbrent committed:
finish vault tests

tbrent committed:
fix Ownable bug with accepting ownership

tbrent committed:
add completed ownable tests

fiddlemath committed:
Local copy of pull review checklist

fiddlemath committed:
A snapshot

tbrent committed:
remove changeName from RSV

tbrent committed:
remove freezing/wiping RSV functionality

tbrent committed:
Merge branch 'working' of github.com:reserve-protocol/rsv-beta into working

tbrent committed:
remove whitelisting + add manager tests

tbrent committed:
fix typos in basket

tbrent committed:
complete refactor; compiles

tbrent committed:
fix new weight proposal bug

fiddlemath committed:
Draft of manager with new basket

fiddlemath committed:
Somehwat cleaner build messages from generate.go

fiddlemath committed:
Compiling version of proposal redesign

fiddlemath committed:
Merge branch 'refactor-proposal-and-basket' of github.com:reserve-protocol/rsv-beta into refactor-proposal-and-basket

fiddlemath committed:
Post-merge fixes to reach compilation

fiddlemath committed:
Merge branch 'working' into refactor-proposal-and-basket

fiddlemath committed:
Merge pull request #6 from reserve-protocol/refactor-proposal-and-basket

Refactor proposal and basket

fiddlemath committed:
Just tweaking some comments

fiddlemath committed:
Cleaning up most remaining TODOs.

The only remaining TODO in the contract comments is to think very clearly about rounding error,
which is a serious task rather than a quick little fix.

fiddlemath committed:
Nitpicky comment edit

tbrent committed:
remove redundant type cast

tbrent committed:
remove outdated TODO comment

tbrent committed:
add comments for proposal contracts

tbrent committed:
remove last instance of unnecessary underscore names

tbrent committed:
restore tests to where they were before

tbrent committed:
fix bug with manager not being able to pass its first proposal

tbrent committed:
combine executeProposal events

tbrent committed:
manager first proposal test pass

tbrent committed:
only we can execute proposals

tbrent committed:
small comment update

tbrent committed:
comment tweaks

tbrent committed:
undo accidentally checked in merge characters

fiddlemath committed:
Handle rounding in RSV weighting.

tbrent committed:
basket tests

tbrent committed:
Merge branch 'working' of github.com:reserve-protocol/rsv-beta into working

fiddlemath committed:
Draft of handling rounding in SwapProposals

fiddlemath committed:
Safemath for mod; remove vault reference from proposal code

fiddlemath committed:
Fixed up rounding a bit, and reordered Manager

tbrent committed:
progress on manager test and init proposal test

fiddlemath committed:
Changed isFullyCollateralized to need no rounding and be easier to reason with.

fiddlemath committed:
Changed weights to be 36-digit fixed point instead of 18-digit fixed point.

tbrent committed:
fix small comment typo

tbrent committed:
add BasketCreated event to proposals

fiddlemath committed:
Improve comments to present correct units and dimensional analysis.

tbrent committed:
rename proposal event

tbrent committed:
check-in confusing failing test

fiddlemath committed:
Do not override proposer in SwapProposal

fiddlemath committed:
Remove comment indicating spurious uncertainty!

fiddlemath committed:
Clean up Basket header comment

fiddlemath committed:
Add Basket::constructor comment

fiddlemath committed:
Subtract 1 from token amount in SwapProposal, to avoid overspending upon execution.

fiddlemath committed:
Make Manager.issue and Manager.redeeem external instead of public.

fiddlemath reviewed (commented):
On solc versions:

In their most recent release of Slither, Trail Of Bits recommends v0.5.3, claiming it's, roughly, old enough to be trusted, and new enough to have critical security updates. Slither also suggests yet mroe strongly, that we use the same version of Solidity everywhere -- say, replacing the `pragma`s with `pragma 0.5.3` instead of `pragma ^0.5.x`.

fiddlemath commented on contracts/ownership/Ownable.sol:
```diff
@@ -0,0 +1,72 @@
+     * @dev Accepts ownership of the contract.
+     */
+    function acceptOwnership() external {
+        require(_nominatedOwner == _msgSender() || _owner == _msgSender(), "unauthorized");
```
Allowing `_owner == _msgSender()` obviates the security purpose of `acceptOwnership`. Do we _need_ this?

tbrent replied:
No I meant to take this out and forgot.

fiddlemath replied:
done

fiddlemath commented on contracts/rsv/Reserve.sol:
```diff
@@ -0,0 +1,293 @@
+    /// This will break this contract, so only do it if you're
+    /// abandoning this contract, e.g., for an upgrade.
+    function transferEternalStorage(address newOwner) external onlyOwner {
+        data.transferOwnership(newOwner);
```
```suggestion
        require(paused);
        data.transferOwnership(newOwner);
```

fiddlemath commented on contracts/rsv/Reserve.sol:
```diff
@@ -0,0 +1,293 @@
+ * @title An interface representing a contract that calculates transaction fees
+ */
+ interface ITXFee {
+    function calculateFee(address, uint256) external returns (uint256);
```
Why does `calculateFee` take just the `from` address, instead of both `from` and `to`?

Also, separately, we should name the parameters here.

tbrent replied:
That seems right.

fiddlemath replied:
done

fiddlemath commented on contracts/rsv/Reserve.sol:
```diff
@@ -0,0 +1,293 @@
+        data.subBalance(from, value);
+        uint256 fee = 0;
+
+        // This is a first-pass implementation and could be totally broken
```
Proposal: when someone executes a transfer of `value` with a nonzero fee, the recipient should get all of `value`, and the fee should _also_ come out of `from`'s account. (And if they can't afford that fee, it'll revert.) That would make it a lot easier to make intended payments, though it is also, admittedly, not the usual way this works with, say, credit cards. Perhaps we should aim to do *better* than them?

tbrent replied:
I'm not sure this is better. If I want to send my entire balance to someone, I don't want to have to calculate what tx fee tier I'm in, calculate the expected nominal amount, and subtract that from the total transfer amount. It's too much work. I expect if we go this route people will not like it.

fiddlemath replied:
Sure, but how often do you want to send your entire balance, instead of to send a little bit?

... ok, actually kind of often. Hm. Seems like these might be two pretty separate use cases; maybe they need two separate external functions. Will leave it open for now.

fiddlemath commented on contracts/rsv/ReserveEternalStorage.sol:
```diff
@@ -0,0 +1,105 @@
+
+
+
+    // ===== frozenTime =====
```
We can drop `frozenTime` entirely if we're dropping freezing. (If we *must*, later, we can use a separate contract to store `frozenTime`.)

fiddlemath replied:
I went ahead and did this. but I still wanted to note it explicitly. @tbrent -- thumbs up or fight me!

fiddlemath commented on contracts/test/BasicOwnable.sol:
```diff
@@ -0,0 +1,8 @@
+pragma solidity ^0.5.8;
```
Does this need to exist at all? `Ownable` should be itself not-abstract.

tbrent replied:
I thought I tried to deploy an `Ownable` instance and couldn't, but looking at it now I can't see how that would be the case. We should be able to remove `BasicOwnable`

fiddlemath commented on tests/Makefile:
```diff
@@ -0,0 +1,56 @@
+DOCKER?=docker
```
Huh. I'm assuming this just isn't getting used at all right now, and can be deleted. Right @tbrent ?

fiddlemath commented on tests/README.md:
```diff
@@ -0,0 +1,10 @@
+tests
```
Likewise, this entire file is also probably not up to date, or useful. Yes? @tbrent ?

fiddlemath committed:
Cleanup Vault by removing unused stuff, and inlining a modifier used only once.

fiddlemath committed:
Add an explicit uint type conversion. Pants and suspenders!

fiddlemath committed:
RSV contract review improvements:

- Bring comments in line with rest of contracts
- Drop frozenTime
- Reorder some things to be easier to read
- Change TXFee handling slightly

fiddlemath committed:
slither triage file.

Handled correctly, this might make it much easier to use slither to double-check needed properties.
On the other hand, maybe you want to make your own choices about what warnings to attend.

fiddlemath committed:
Remove stale README and Makefile

tbrent committed:
Update contracts/rsv/Reserve.sol

Co-Authored-By: Matt Elder <fiddlemath@users.noreply.github.com>

fiddlemath committed:
Do not allow owner to call acceptOwnership

fiddlemath committed:
tx fee helper's calculateFee should take from and to addresses.

fiddlemath committed:
Use solc 0.5.7 throughout this project.

Why 0.5.7? It's the oldest version that:
- supports all the features our code actually uses (the extcodehash op)
- was not shortly followed by bugfixes (like 0.5.5 and 0.5.6 were)

fiddlemath commented:
> On solc versions:
> 
> In their most recent release of Slither, Trail Of Bits recommends v0.5.3, claiming it's, roughly, old enough to be trusted, and new enough to have critical security updates. Slither also suggests yet mroe strongly, that we use the same version of Solidity everywhere -- say, replacing the `pragma`s with `pragma 0.5.3` instead of `pragma ^0.5.x`.

Commit 47f6d3d changes everything to use solc v0.5.7.
